docs/source/python/extending_types.rst

```diff
@@ -0,0 +1,115 @@
+.. KIND, either express or implied.  See the License for the
+.. specific language governing permissions and limitations
+.. under the License.
+
+.. currentmodule:: pyarrow
+.. _extending_types:
+
+Extending pyarrow
+=================
+
+Defining extension types ("user-defined types")
+-----------------------------------------------
+
+Arrow has the notion of extension types in the metadata specification as a
+possiblity to extend the built-in types. This is done by annotating any of the
+built-in Arrow logical types (the "storage type") with a custom type name and
+optional serialized representation ("ARROW:extension:name" and
+"ARROW:extension:metadata" keys in the Field’s custom_metadata of an IPC
+message).
+See the :ref:`format_metadata_extension_types` section of the metadata
+specification for more details.
+
+Pyarrow allows you to define such extension types from Python.
+
+There are currently two ways:
+
+* Subclassing :class:`ExtensionType`: the (de)serialization is based on pickle.
+  This is a good option for an extension type that is only used from Python.
+* Subclassing :class:`GenericExtensionType`: this allows to give a custom
+  Python-independent name and serialized metadata, that can potentially be
+  recognized by other (non-Python) Arrow implementations such as PySpark.
+
+
+For example, we could define a custom UUID type for 128-bit numbers which can
+be represented as ``FixedSizeBinary`` type with 16 bytes.
+Using the first approach, we create a `UuidType` subclass, and implement the
+``__reduce__`` dunder method to ensure the class can be properly pickled::
+
+    class UuidType(pa.ExtensionType):
+
+        def __init__(self):
+            pa.ExtensionType.__init__(self, pa.binary(16))
+
+        def __reduce__(self):
+            return UuidType, ()
+
+This can now be used to create arrays and tables holding the extension type::
+
+    >>> uuid_type = UuidType()
+    >>> uuid_type.extension_name
+    'arrow.py_extension_type'
+    >>> uuid_type.storage_type
+    FixedSizeBinaryType(fixed_size_binary[16])
+
+    >>> import uuid
+    >>> storage_array = pa.array([uuid.uuid4().bytes for _ in range(4)], pa.binary(16))
+    >>> arr = pa.ExtensionArray.from_storage(uuid_type, storage_array)
+    >>> arr
+    <pyarrow.lib.ExtensionArray object at 0x7f75c2f300a0>
+    [
+      A6861959108644B797664AEEE686B682,
+      718747F48E5F4058A7261E2B6B228BE8,
+      7FE201227D624D96A5CD8639DEF2A68B,
+      C6CA8C7F95744BFD9462A40B3F57A86C
+    ]
+
+This array can be included in RecordBatches, sent over IPC and received in
+another Python process. The custom UUID type will be preserved there, as long
+as the definition of the class is available (the type can be unpickled).
+
+We can define the same type using the other option::
+
+    class UuidType(pa.GenericExtensionType):
+
+        def __init__(self):
+            pa.GenericExtensionType.__init__(self, pa.binary(16), "my_package.uuid")
+
+        def __arrow_ext_serialize__(self):
```

Comment:
I think it's fine to keep them as-is. It hints that they're not meant to be called directly by the user; rather, Arrow is calling them.

Reply:
I agree with @pitrou here. For me, it seems similar to methods like `__eq__` etc that you can define on your class, as those are also not meant to be called directly by the end-user, but are called by low-level code.

Reply:
The name might be misleading, but as far as I understand, this is not meant to be a proper standalone serialization method. 
With the current metadata specification, the serialized bytes don't need to completely describe the type (and thus are not necessarily fully descriptive to result in the type when deserialized), but only in combination with a type name and a registry of types. 

The result of the "serialize" method is stored in the "ARROW:extension:metadata" key of the field's metadata. So it is more "serialized form of the type's metadata", than actually a "serialized form of the type".

Of course, it is not really specified (as far as I know), how this metadata field needs to be used, and it could also hold a serialization of the full type. Like the pickle-based PyExtensionType does, but that is specific to that specific implementation.